source/slang/slang-ir-diff-jvp.cpp

```diff
@@ -3,10 +3,303 @@
+            // like float3 or float3x3
+            // 
+            auto resultType = arith->getDataType();
+            switch(arith->getOp())
+            {
+            case kIROp_Add:
+                return builder->emitAdd(resultType, leftD, rightD);
+            case kIROp_Mul:
+                return builder->emitAdd(resultType,
+                    builder->emitMul(resultType, leftD, rightP),
+                    builder->emitMul(resultType, leftP, rightD));
+            case kIROp_Sub:
+                return builder->emitSub(resultType, leftD, rightD);
+            case kIROp_Div:
+                return builder->emitDiv(resultType, 
+                    builder->emitSub(
+                        resultType,
+                        builder->emitMul(resultType, leftD, rightP),
+                        builder->emitMul(resultType, leftP, rightD)),
+                    builder->emitMul(
+                        rightP->getDataType(), rightP, rightP
+                    ));
+            default:
+                SLANG_UNEXPECTED("Attempting to differentiate unsupported arithmetic");
+            }
+        }
+
+        return nullptr;
+    }
+
+    IRInst* differentiateLoad(IRBuilder* builder, IRLoad* loadP)
+    {
+        if (auto varP = as<IRVar>(loadP->getPtr()))
+        {   
+            // If the loaded parameter has a differential version, 
+            // emit a load instruction for the differential parameter.
+            // Otherwise, emit nothing since there's nothing to load.
+            // 
+            if (auto varD = as<IRVar>(getDifferentialInst(varP)))
+            {
+                IRLoad* loadD = as<IRLoad>(builder->emitLoad(varD));
+                SLANG_ASSERT(loadD);
+                return loadD;
+            }
+            return nullptr;
+        }
+
+        SLANG_UNEXPECTED("Attempting to differentiate an unsupported load instruction");
+    }
+
+    IRInst* differentiateStore(IRBuilder* builder, IRStore* storeP)
+    {
+        IRInst* storeLocation = storeP->getPtr();
+        IRInst* storeVal = storeP->getVal();
+        if (auto destParam = as<IRVar>(storeLocation))
+        {   
+            // If the stored value has a differential version, 
+            // emit a store instruction for the differential parameter.
+            // Otherwise, emit nothing since there's nothing to load.
+            // 
+            IRInst* storeValD = getDifferentialInst(storeVal);
+            IRVar*  storeLocationD = as<IRVar>(getDifferentialInst(destParam));
+            if (storeValD && storeLocationD)
+            {
+                IRStore* storeD = as<IRStore>(
+                    builder->emitStore(storeLocationD, storeValD));
+                SLANG_ASSERT(storeD);
+                return storeD;
+            }
+            return nullptr;
+        }
+        
+        SLANG_UNEXPECTED("Attempting to differentiate an unsupported store instruction");
```

Comment:
We realistically should be checking this in the front-end to make sure that doesn't happen here. But for now we can just issue a diagnostic for unsupported case in the IR pass. Let's not use SLANG_ASSERT or SLANG_UNEXPECTED in our IR passes if possible.